src/d3a/models/strategy/external_strategy.py

```diff
@@ -0,0 +1,125 @@
+        # if "offer" in data_dict and data_dict["offer"] is not None:
+        #     if isinstance(data_dict["offer_or_id"], str):
+        #         data_dict["offer_or_id"] = offer_from_JSON_string(data_dict["offer_or_id"])
+
+        return data_dict
+
+    @staticmethod
+    def _serialize_offer_list(offer_list):
+        return json.dumps([offer.to_JSON_string() for offer in offer_list])
+
+    @staticmethod
+    def _serialize_offer_dict(offer_dict):
+        return json.dumps({k: v.to_JSON_string() for k, v in offer_dict.items()})
+
+    def _offer_lists(self, payload):
+        try:
+            return_data = self._serialize_offer_dict(self.area.next_market.offers)
+            print(return_data)
+            self.publish(f"{self.area.parent.slug}/{self.area.slug}/offers/response",
+                         {"status": "ready", "offer_list": return_data})
+        except Exception as e:
+            self.publish(f"{self.area.parent.slug}/{self.area.slug}/offers/response",
+                         {"status": "error",  "exception": str(type(e)),
+                          "error_message": str(e)})
+
+    def _accept_offer(self, payload):
+        try:
+            trade = self.area.next_market.accept_offer(**self._parse_payload(payload))
+            self.publish(f"{self.area.parent.slug}/{self.area.slug}/accept_offer/response",
+                         {"status": "ready", "trade": trade.to_JSON_string()})
+        except Exception as e:
+            self.publish(f"{self.area.parent.slug}/{self.area.slug}/accept_offer/response",
+                         {"status": "error",  "exception": str(type(e)),
+                          "error_message": str(e)})
+
+    def _offer(self, payload):
+        try:
+            offer = self.area.next_market.offer(**self._parse_payload(payload))
+            self.publish(f"{self.area.parent.slug}/{self.area.slug}/offer/response",
+                         {"status": "ready", "offer": offer.to_JSON_string()})
+        except Exception as e:
+            self.publish(f"{self.area.parent.slug}/{self.area.slug}/offer/response",
+                         {"status": "error",  "exception": str(type(e)),
+                          "error_message": str(e)})
+
+    def _delete_offer(self, payload):
+        try:
+            self.area.next_market.delete_offer(**self._parse_payload(payload))
+            self.publish(f"{self.area.parent.slug}/{self.area.slug}/delete_offer/response",
+                         {"status": "ready"})
+        except Exception as e:
+            self.publish(f"{self.area.parent.slug}/{self.area.slug}/delete_offer/response",
+                         {"status": "error", "exception": str(type(e)),
+                          "error_message": str(e)})
+
+
+class ExternalStrategy(BaseStrategy):
+    def __init__(self, area):
+        super().__init__()
+        self.redis = RedisExternalConnection(area)
+
+    def event_market_cycle(self):
+        pass
```

Comment:
Shouldn't there also be `event_tick`  method here in case user wants to update his/her offer/bid?

Reply:
Not really, because the user does not care about ticks in this case. The user can place offers to the market even in-between ticks, since the redis connection is running on a thread. Therefore, the user does not need to get notified when to place offers, one can put the offers at will, without waiting for a tick.